pkg/network/encoding/kafka.go

```diff
@@ -6,140 +6,111 @@
-// TODO: we should probably revist/get rid of this if we ever replace socket
-// filters by kprobes, since in that case we would have access to PIDs, and
-// could incorporate that information in the `types.ConnectionKey` struct.
-type kafkaAggregationWrapper struct {
-	*model.DataStreamsAggregations
-
-	// we keep track of the source and destination ports of the first
-	// `ConnectionStats` to claim this `DataStreamsAggregations` object
-	sport, dport uint16
+type kafkaEncoder struct {
+	byConnection *USMConnectionIndex[kafka.Key, *kafka.RequestStat]
+
+	// cached object
+	aggregations *model.DataStreamsAggregations
 }
 
-func (a *kafkaAggregationWrapper) ValueFor(c network.ConnectionStats) *model.DataStreamsAggregations {
-	if a == nil {
+func newKafkaEncoder(payload *network.Connections) *kafkaEncoder {
+	if len(payload.Kafka) == 0 {
 		return nil
 	}
 
-	if a.sport == 0 && a.dport == 0 {
-		// This is the first time a ConnectionStats claim this aggregation. In
-		// this case we return the value and save the source and destination
-		// ports
-		a.sport = c.SPort
-		a.dport = c.DPort
-		return a.DataStreamsAggregations
+	return &kafkaEncoder{
+		aggregations: &model.DataStreamsAggregations{
+			// It's not that important to get the initial size of this slice
+			// right because we're re-using it multiple times and should quickly
+			// converge to a final size after a few calls to
+			// `GetKafkaAggregations`
+			KafkaAggregations: make([]*model.KafkaAggregation, 0, 10),
+		},
+		byConnection: GroupByConnection("kafka", payload.Kafka, func(key kafka.Key) types.ConnectionKey {
+			return key.ConnectionKey
+		}),
 	}
-
-	if c.SPort == a.dport && c.DPort == a.sport {
-		// We have a collision with another `ConnectionStats`, but this is a
-		// legit scenario where we're dealing with the opposite ends of the
-		// same connection, which means both server and client are in the same host.
-		// In this particular case it is correct to have both connections
-		// (client:server and server:client) referencing the same Kafka data.
-		return a.DataStreamsAggregations
-	}
-
-	// Return nil otherwise. This is to prevent multiple `ConnectionStats` with
-	// exactly the same source and destination addresses but different PIDs to
-	// "bind" to the same DataStreamsAggregations object, which would result in a
-	// overcount problem. (Note that this is due to the fact that
-	// `types.ConnectionKey` doesn't have a PID field.) This happens mostly in the
-	// context of pre-fork web servers, where multiple worker proceses share the
-	// same socket
-	return nil
 }
 
-func newKafkaEncoder(payload *network.Connections) *kafkaEncoder {
-	if len(payload.Kafka) == 0 {
+func (e *kafkaEncoder) GetKafkaAggregations(c network.ConnectionStats) []byte {
+	if e == nil {
 		return nil
 	}
 
-	encoder := &kafkaEncoder{
-		aggregations: make(map[types.ConnectionKey]*kafkaAggregationWrapper, len(payload.Conns)),
+	connectionData := e.byConnection.Find(c)
```

Comment:
Isn't this going to find all the connections since the encoder is built from c ?

Reply:
The encoder wasn't built from `c`, but from the kafka payloads (based on USM connections)
now we're trying to find a matching npm connection